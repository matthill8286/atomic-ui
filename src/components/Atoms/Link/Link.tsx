import React from 'react'
import { Link as RouterLink } from 'react-router-dom'
import { smoothScrollWithEvent } from '@/utils'
import { LinkProps, StyledLinkProps } from './Link.interface'
import {
  StyledIconWrapper,
  StyledLinkAnchor,
  StyledLinkButton,
  StyledLinkRouter,
  Text,
} from './Link.styled'

export const Link: React.FC<LinkProps> = ({
  bold = false,
  branded = false,
  children,
  className,
  cursor,
  color = 'grey4',
  decorationColor,
  disabled = false,
  fontSize,
  href,
  iconLeft,
  iconRight,
  inline = false,
  lineHeight,
  onClick,
  onMouseDown,
  scale = 'small',
  scrollOffset = 0,
  smooth = false,
  target = '_self',
  title,
  to,
  underline = true,
  isGrouped = false,
  ariaLabel,
  ...otherProps
}): JSX.Element => {
  const isSmall = scale === 'small'
  const isButton = onClick && !href && !to
  const isRouterLink = to

  const styleProps: StyledLinkProps = {
    color: decorationColor || color,
    cursor,
    hasIconLeft: !!iconLeft,
    hasIconRight: !!iconRight,
    isBold: bold,
    isDisabled: disabled,
    isInline: inline,
    isLarge: scale === 'large',
    isSmall,
    isUnderlined: underline,
    isGrouped,
  }
  let onClickWrapper

  if (smooth) {
    onClickWrapper = (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => {
      try {
        smoothScrollWithEvent(event, scrollOffset)
      } catch (e) {
        //failsafe
      }
      if ('function' == typeof onClick) {
        return onClick(event)
      }
    }
  } else {
    onClickWrapper = onClick
  }

  const getChildren = (): JSX.Element => (
    <>
      {iconLeft && <StyledIconWrapper {...styleProps}>{iconLeft}</StyledIconWrapper>}
      {typeof children === 'string' ? (
        <Text
          tag="span"
          fontSize={fontSize || (isSmall ? 'xs' : 'sm')}
          lineHeight={lineHeight}
          bold={bold}
          color={color}
          isBranded={branded}
          isInline={inline}
          aria-label={ariaLabel}>
          {children}
        </Text>
      ) : (
        children
      )}
      {iconRight && <StyledIconWrapper {...styleProps}>{iconRight}</StyledIconWrapper>}
    </>
  )

  const linkType = (): JSX.Element => {
    if (isButton) {
      return (
        <StyledLinkButton
          className={className}
          {...(disabled ? {} : { onClick: onClickWrapper })}
          {...styleProps}
          {...otherProps}>
          {getChildren()}
        </StyledLinkButton>
      )
    }

    if (isRouterLink) {
      return (
        <StyledLinkRouter
          onClick={onClickWrapper}
          to={to}
          className={className}
          {...styleProps}
          data-test="ewb-router-link">
          {getChildren()}
        </StyledLinkRouter>
      )
    }

    return (
      <StyledLinkAnchor
        className={className}
        target={target}
        rel={target === '_blank' ? 'noopener noreferrer' : otherProps.rel}
        {...(disabled ? {} : { href, onClick: onClickWrapper })}
        {...styleProps}
        {...otherProps}>
        {getChildren()}
      </StyledLinkAnchor>
    )
  }

  return linkType()
}
