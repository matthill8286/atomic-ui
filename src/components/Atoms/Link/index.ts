export * from './Link'
export * from './Link.interface'
