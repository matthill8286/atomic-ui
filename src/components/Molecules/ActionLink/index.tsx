export * from './ActionLink.interface'
export * from './ActionLink'
export * from './GroupedActionLinks'
