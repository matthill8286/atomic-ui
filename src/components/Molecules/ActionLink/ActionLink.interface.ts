import { MouseEvent } from 'react'
import { LinkProps } from '../../Atoms/Link/Link.interface'
import { TranslatedTextType } from '../../../types'

export interface ActionLinkProps extends LinkProps {
  children: TranslatedTextType
}

export interface QuickAction {
  actionLabel: string
  href?: string
  to?: string
  target?: string
  onClick?: (e: MouseEvent<HTMLElement>) => void
}
