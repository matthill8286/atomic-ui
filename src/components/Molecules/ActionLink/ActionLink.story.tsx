import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import React from 'react'
import { ActionLink } from './ActionLink'
import { ActionLinkProps } from './ActionLink.interface'

const actionLinkChildren = 'Action link'
const Template = (args: ActionLinkProps) => <ActionLink {...args} />

storiesOf('Design System/Molecules/Action Link', module).add('Single', () => (
  <Template>{actionLinkChildren}</Template>
))

storiesOf('Design System/Molecules/Action Link', module).add('On Click', () => (
  <Template onClick={action('Action Link clicked')}>{actionLinkChildren}</Template>
))
