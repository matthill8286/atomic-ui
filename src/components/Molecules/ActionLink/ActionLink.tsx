import React, { FC } from 'react'
import { Link, LinkProps } from '@/components/Atoms/Link'
import { ActionLinkProps } from './ActionLink.interface'

export const ActionLink: FC<ActionLinkProps> = ({
  children,
  bold = false,
  iconLeft,
  ...baseLinkProps
}) => {
  const linkProps: LinkProps = {
    ...baseLinkProps,
    iconLeft,
    underline: false,
    scale: 'large',
    bold,
    isGrouped: true,
  }

  return <Link {...linkProps}>{children}</Link>
}
