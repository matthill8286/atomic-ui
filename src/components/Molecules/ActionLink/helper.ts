import { QuickAction } from './ActionLink.interface'

export const getActionLinkType = (quickAction: QuickAction) => {
  if (quickAction.href) {
    return {
      href: quickAction.href,
    }
  } else if (quickAction.to) {
    return {
      to: quickAction.to,
    }
  }
}
