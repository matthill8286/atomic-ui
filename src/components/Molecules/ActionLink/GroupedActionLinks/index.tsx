import React, { FC } from 'react'
import { QuickAction } from '../ActionLink.interface'
import { getActionLinkType } from '../helper'
import { ActionLink } from '../'

export interface GroupedActionLinksProps {
  links: QuickAction[]
}

export const GroupedActionLinks: FC<GroupedActionLinksProps> = ({ links }) => {
  return (
    <>
      {links.map((quickAction, index) => {
        const linkProp = getActionLinkType(quickAction)
        return (
          <ActionLink
            key={index}
            target={quickAction.target}
            onClick={quickAction.onClick}
            {...linkProp}>
            {quickAction.actionLabel}
          </ActionLink>
        )
      })}
    </>
  )
}
