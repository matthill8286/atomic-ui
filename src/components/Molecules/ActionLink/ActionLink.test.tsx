import React from 'react'
import { Link } from '@/components/Atoms/Link'
import { mountWithTheme, renderWithTheme } from '@/testRenderer'
import { ActionLink } from './ActionLink'
import { ActionLinkProps } from './ActionLink.interface'

describe('<ActionLink /> component', () => {
  const props: ActionLinkProps = {
    children: 'Click me',
  }

  const defaultActionLink = <ActionLink {...props} />

  test('should render a <Link /> component', () => {
    const actionLink = renderWithTheme(defaultActionLink)
    const wrapper = mountWithTheme(defaultActionLink)
    expect(wrapper.find(Link)).toHaveLength(1)
    expect(actionLink).toMatchSnapshot()
  })

  test('should trigger onClick', () => {
    const onClick = jest.fn()
    const wrapper = mountWithTheme(<ActionLink {...props} onClick={onClick} />)

    wrapper.find(Link).simulate('click')
    expect(onClick).toHaveBeenCalled()
  })
})
