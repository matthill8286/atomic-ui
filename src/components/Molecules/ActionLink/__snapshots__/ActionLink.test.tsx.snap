// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionLink /> component should render a <Link /> component 1`] = `
.c1 {
  color: #999999;
  font-size: 16px;
  line-height: 24px;
  font-family: "Oswald","Raleway",Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c0::after {
  content: '';
  position: absolute;
  display: none;
  bottom: -2px;
  height: 1px;
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c2 {
  display: block;
  color: #999999;
}

<a
  className="c0"
  color="grey4"
  target="_self"
>
  <span
    className="c1 c2"
    color="grey4"
    fontSize="sm"
  >
    Click me
  </span>
</a>
`;
